AIP 39: Documentation

uranusjr opened the pull request:
~Very much a work in progress.~ Ready for review now.

I have very low confidence on my ability writing comprehensive prose 🙂

uranusjr committed:
Mention data interval and add timetable how-to

eladkal commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,31 @@ Cron Presets
+its data interval would start at midnight of each day, and end at midnight of
+the next day.
+
+A DAG run happens *after* its associated data interval has ended, to ensure the
```
```suggestion
A DAG run created *after* its associated data interval has ended, to ensure the
```

jedcunningham replied:
```suggestion
A DAG run is created *after* its associated data interval has ended, to ensure the
```

"happens" is fine, otherwise it should be "is created". The latter might be a better choice, but really either work.

uranusjr replied:
I was trying to avoid the passive “is created” because the data interval’s end does not technically trigger the run’s creation, but merely triggers a request to create a run. I’m not sure whether this distinction is meaningful to the user (from an end user’s perspective, this means the run won’t always be created immediately after the data interval ends, but only sometime later). I switched to use “is scheduled” here instead, but other suggestions are still welcomed.

eladkal commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,31 @@ Cron Presets
+when the DAG is actually executed. Similarly, since the ``start_date`` argument
+for the DAG and its tasks points to the same logical date, a run will only
+be created after that data interval ends. So a DAG with ``@daily`` schedule and
+``start_date`` of 2020-01-01, for example, will not be created until 2020-01-02.
```
What is the idea we want to deliver here? (It feels like the same information given in rows 60-63)

uranusjr replied:
The previous paragraph tries to establish the idea of data intervals in general, and this paragraph explains the concept underlies almost all of Airflow’s date-related configurations, including (the most commonly misunderstood) `start_date`. The intuitive interpretation from the argument name is my DAG should *start running* after this date, while in fact only the DAG’s *first data interval* starts after it. I’ve reworded both paragraphs a bit to make them read more different.

jedcunningham commented on docs/apache-airflow/concepts/dags.rst:
```diff
@@ -148,14 +148,24 @@ The ``schedule_interval`` argument takes any value that is a valid `Crontab <htt
 As an example of why this is useful, consider writing a DAG that processes a daily set of experimental data. It's been rewritten, and you want to run it on the previous 3 months of data - no problem, since Airflow can *backfill* the DAG and run copies of it for every day in those previous 3 months, all at once.
 
-Those DAG Runs will all have been started on the same actual day, but their ``execution_date`` values will cover those last 3 months, and that's what all the tasks, operators and sensors inside the DAG look at when they run.
+Those DAG Runs will all have been started on the same actual day, but their data intervals will cover those last 3 months, and that's what all the tasks, operators and sensors inside the DAG look at when they run.
```
```suggestion
Those DAG Runs will all have been started on the same actual day, but each DAG run will have a data interval covering a single day in that 3 month period, and that data interval is all the tasks, operators and sensors inside the DAG look at when they run.
```

I think this is more clear, but I'm not hellbent on this one.

uranusjr replied:
Definitely clearer than the previous wording!

jedcunningham commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,31 @@ Cron Presets
+
+Each DAG run in Airflow has an assigned "data interval" that represents the time
+range it operates in. For a DAG scheduled with ``@daily``, for example, each of
+its data interval would start at midnight of each day, and end at midnight of
```
```suggestion
its data interval would start at midnight of each day and end at midnight of
```

jedcunningham commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,31 @@ Cron Presets
+the next day.
+
+A DAG run happens *after* its associated data interval has ended, to ensure the
+run is able to collect all the actual data within the time period. Therefore, a
```
```suggestion
run is able to collect all the data within the time period. Therefore, a
```

jedcunningham commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,31 @@ Cron Presets
+2020-01-01 has ended, i.e. 2020-01-02 onwards.
+
+All dates in Airflow are tied to the data interval concept in some way. The
+"logical date" (also called ``execution_date`` from previous Airflow version)
```
```suggestion
"logical date" (also called ``execution_date`` in Airflow versions before 2.2)
```

jedcunningham commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,31 @@ Cron Presets
+when the DAG is actually executed. Similarly, since the ``start_date`` argument
+for the DAG and its tasks points to the same logical date, a run will only
+be created after that data interval ends. So a DAG with ``@daily`` schedule and
+``start_date`` of 2020-01-01, for example, will not be created until 2020-01-02.
```
```suggestion
be created after that data interval ends. So a DAG with a ``@daily`` schedule and
a ``start_date`` of 2020-01-01, for example, will not be created until 2020-01-02.
```

uranusjr replied:
I decided to remove this example since it is too similar to the one in the previous paragraph and could be more confusing than clarifying.

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+* Rolling windows, or overlapping data intervals. For example, one may want to
+  have a run each day, but make each run cover the period of the previous seven
+  days. It is possible to "hack" this with a cron expression, but a custom data
+  interval would task specification more natural.
```
```suggestion
  interval would be a more natural representation.
```

Or

```suggestion
  interval would make the task specification more natural.
```

uranusjr replied:
I chose the first one

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+
+
+For our example, let's say a company may want to run a job after each weekday,
+to process data collected during the work day. The first intuitively answer
```
```suggestion
For our example, let's say a company wants to run a job after each weekday
to process data collected during the work day. The first intuitive answer
```

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+midnight Friday to midnight *Monday*.
+
+This is, therefore, a case of the "holes" category; the intended schedule should
+leave the two weekend days. What we want is:
```
```suggestion
This is, therefore, an example in the "holes" category above; the intended schedule should
not include the two weekend days. What we want is:
```

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+leave the two weekend days. What we want is:
+
+* Schedule a run for each Monday, Tuesday, Wednesday, Thursday, and Friday. The
+  run's data interval would cover from the midnight of each day, to the midnight
```
```suggestion
  run's data interval would cover from midnight of each day, to midnight
```

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+
+* Schedule a run for each Monday, Tuesday, Wednesday, Thursday, and Friday. The
+  run's data interval would cover from the midnight of each day, to the midnight
+  of the next day.
```
```suggestion
  of the next day (e.g. 2021-01-01 00:00:00 to 2021-01-02 00:00:00).
```

Maybe add an example?

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+* Data intervals with "holes" between. (Instead of continous, as both the cron
+  expression and ``timedelta`` schedules represent.)
+* Run tasks on different times each day. For example, an astronomer may find it
+  useful to run a task on each sunset, to process data collected from the
```
```suggestion
* Run tasks at different times each day. For example, an astronomer may find it
  useful to run a task at sunset to process data collected from the
```

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,63 @@
+to process data collected during the work day. The first intuitively answer
+to this would be ``schedule_interval="0 0 * * 1-5"`` (midnight on Monday to
+Friday), but this means data collected on Friday will *not* be processed right
+after Friday, but on the next Monday, and that run's interval would be from
```
```suggestion
after Friday ends, but on the next Monday, and that run's interval would be from
```

ashb commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,33 @@ Cron Presets
+A DAG run is scheduled *after* its associated data interval has ended, to ensure
+the run is able to collect all the data within the time period. Therefore, a run
+covering the data period of 2020-01-01 will not start to run until 2020-01-01
+has ended, i.e. after 2020-01-02 00:00:00.
```
This doesn't _have_ to be the case, it's up to the timetable to decide when it's schedulable isn't it?

uranusjr replied:
Technically yes; I wrote it like this to avoid going into too much “well except” stuff this early in the documentation. It’d definitely be best if we can explain this default “DAG skips one interval” behaviour in another way.

uranusjr replied:
I made the wording less certain on this to acount for the custom timetable edge cases. I don’t think running a DAG before its data interval ends makes enough general sense to complicate the explaination, so this is done with some “usually” and “generally”s.

ashb commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+  expression and ``timedelta`` schedules represent.)
+* Run tasks at different times each day. For example, an astronomer may find it
+  useful to run a task at sunset to process data collected from the previous
+  sunlight period.
```
```suggestion
  useful to run a task at dawn to process data collected from the previous
  night-time period.
```

Seems more realistic :)

ashb commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+
+* ``earliest``: The earliest time the DAG may be scheduled. This is a
+  calculated ``pendulum.DateTime`` from all the ``start_date`` arguments from
+  the DAG and its tasks, or ``Non`` of there are no ``start_date`` arguments
```
```suggestion
  the DAG and its tasks, or ``None`` if there are no ``start_date`` arguments
```

ashb commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+* ``data_interval``: A :class:`~airflow.timetables.base.DataInterval` instance
+  like ``infer_data_interval``'s return value. This describes the next run's
+  data interval.
+* ``run_after: A ``pendulum.DateTime`` instance that tells the scheduler when
```
```suggestion
* ``run_after``: A ``pendulum.DateTime`` instance that tells the scheduler when
```

ashb commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+----------------------------------------------
+
+Sometimes we need to pass some run-time arguments to the timetable. Continuing
+with out ``AfterWorkdayTimetable``, maybe we may have DAGs running on different
```
```suggestion
with our ``AfterWorkdayTimetable`` example, maybe we may have DAGs running on different
```

ashb commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+
+When the DAG is being serialized, ``serialize`` is called to obtain a
+JSON-serializable value. That value is passed to ``deserialize`` when the
+serialized DAG is accessed by the scheduler to reconstruct the timetable.
```
(Looks fine, but this code isn't landed yet, right?)

uranusjr replied:
Yes, I’m anticipating the actual code change in #17414.

ashb replied:
There might be a thing we need to add there about "registering" the class too, but we can add that to the docs when we decide.

ashb reviewed (commented):
This is really clear to me (but then I'm familiar with both the code and the proposal, so I'm not the best to judge.)

uranusjr commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+* ``earliest``: The earliest time the DAG may be scheduled. This is a
+  calculated ``pendulum.DateTime`` from all the ``start_date`` arguments from
+  the DAG and its tasks, or ``None`` oi there are no ``start_date`` arguments
+  found at all.
```
```suggestion
* ``earliest``: The earliest time the DAG may be scheduled. This is a
  ``pendulum.DateTime`` calculated from all the ``start_date`` arguments from
  the DAG and its tasks, or ``None`` if there are no ``start_date`` arguments
  found at all.
```

uranusjr commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+        run_after=run_after,
+    )
+
+But since we typically wan to schedule a run as soon as the data interval ends,
```
```suggestion
But since we typically want to schedule a run as soon as the data interval ends,
```

uranusjr commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,261 @@
+----------------------------------------------
+
+Sometimes we need to pass some run-time arguments to the timetable. Continuing
+with our ``AfterWorkdayTimetable`` example, maybe we may have DAGs running on
```
```suggestion
with our ``AfterWorkdayTimetable`` example, maybe we have DAGs running on
```

jhtimmins commented on airflow/example_dags/example_workday_timetable.py:
```diff
@@ -0,0 +1,92 @@
+                delta = timedelta(days=1)
+            else:  # Week is ending -- skip to next Monday.
+                delta = timedelta(days=(7 - weekday))
+            start = DateTime.combine((last_automated_dagrun + delta).date(), Time.min)
```
Reminder @uranusjr that there is a bug in this example. The next interval should start at a delta from the start of the incoming dagrun, not the dagrun execution time.

uranusjr replied:
Some additional context:

> it takes the execution time (which falls after the interval ends) and then adds 1 day to it to find the next start of the next interval. But that ends up skipping an interval, if I'm not mistaken.

This makes me think maybe we should pass the entire `DagRunInfo` into `next_dagrun_info` instead of just the previous run date. It is of course not too difficult to reverse-infer the data interval from the run date in this example, but it might not be always as easy or even possible in more elaborated data interval logic.

ashb replied:
Yeah this sounds like a good idea

uranusjr replied:
Hmm, on a closer look, the current logic seems to be correct? The default behaviour of `DagRunInfo` actually defines the execution date as the _start_ of the interval, so I think using `execution_date + delta` is correct. See the `logical_date` property defined here:

https://github.com/apache/airflow/blob/873d4d1f1aad214b36218d80000f5175ae284f6f/airflow/timetables/base.py#L85-L91

uranusjr replied:
Regarding also passing in the data interval. This turns out to be more work than I originally anticipated because it means we’ll need to deprecate `DAG.following_schedule()` (which operates on `execution_date` only).  I’ll work on it in the following days.

uranusjr replied:
New `next_dagrun_info` implementation using data interval instead was merged in #18088. I’ve rewritten the example to match the actual implementation.

uranusjr committed:
Document timetable registration via plugin

uranusjr committed:
Typos

uranusjr committed:
Fix imports in example DAG

uranusjr committed:
Rewrite example with new last_automated_dagrun

uranusjr committed:
Remove execution date reference in docs

Not all usages can be removed, unfortunately, since there are still many
usages requiring the value, especially in Python APIs.

This also adds additional values for Sentry tags and Elastic log
filenames based on data interval instead of execution date, so the user
can migrate as soon as possible, even if we have not changed the default.

uranusjr committed:
Airflow's CI is American

uranusjr committed:
More typos

uranusjr committed:
Fix example

uranusjr committed:
Fix sentry tests to account for additional tags

uranusjr committed:
Revert TI.xcom_pull() usage

We need to use execution_date here :/

uranusjr committed:
Fix Elastic tests depending on private function

uranusjr committed:
Make the run_after description a bit vague

A DAG run CAN technically start before its data interval ends, but
that's an edge case we shouldn't cover early in the documentation, so
let's make wording vague to focus on the general cases to make things
less complicated for users reading this for the first time.

uranusjr committed:
Fix example DAG

uranusjr committed:
Inline the example custom timetable DAG

A custom timetable requires plugin initialization, which is not done for
unit tests. But we really don't want to automatically register that
plugin since it makes testing an unregistered plugin impossible. Since
that example DAG is very short and referenced only once anyway, let's
just not bother.

jedcunningham commented on docs/apache-airflow/best-practices.rst:
```diff
@@ -43,21 +43,26 @@ Please follow our guide on :ref:`custom Operators <custom_operator>`.
+  in a task. Someone may update the input data between re-runs, which results in
+  different outputs. A better way is to read the input data from a specific
+  partition. You can use ``data_interval_start`` as a partition. You should
+  follow this partitioning method while writing data in S3/HDFS, as well.
```
```suggestion
  follow this partitioning method while writing data in S3/HDFS as well.
```

uranusjr replied:
Side note, there are a lot of *blah blah blah, as well* usages in the documentation, so it seems like whoever wrote it previously has a particular style. (I’ve removed the comma here.)

jedcunningham commented on docs/apache-airflow/concepts/dags.rst:
```diff
@@ -317,7 +326,7 @@ Depends On Past
 You can also say a task can only run if the *previous* run of the task in the previous DAG Run succeeded. To use this, you just need to set the ``depends_on_past`` argument on your Task to ``True``.
 
-Note that if you are running the DAG at the very start of its life - specifically, that the ``execution_date`` matches the ``start_date`` - then the Task will still run, as there is no previous run to depend on.
+Note that if you are running the DAG at the very start of its life---specifically, its first ever *automated* run---then the Task will still run, as there is no previous run to depend on.
```
```suggestion
Note that if you are running the DAG at the very start of its life (specifically, its first ever *automated* run) then the Task will still run, as there is no previous run to depend on.
```

This looks better to my eyes?

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,298 @@
+or three days if it was on Friday. If there was not a previous scheduled run,
+however, we pick the next workday's midnight after ``restriction.earliest``
+(unless it *is* a workday's midnight; in which case it's used directly).
+``restriction.catchup`` also needs to be considered---if it's ``False``, we
```
```suggestion
``restriction.catchup`` also needs to be considered - if it's ``False``, we
```

uranusjr replied:
Sphinx renders triple-dash as an em-dash by default. This can be disabled, but from what I can tell we currently have it on.

https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-smartquotes

ashb replied:
emdash ftw. Nice typography makes me happy.

jedcunningham replied:
Ah, cool, TIL. Do we want `<space>---<space>` then so we have spaces?

uranusjr replied:
I quickly searched and there doesn’t seem to be an established rule (both are generally accepted), so I’ll defer this to whatever the majority of reviewers think.

ashb replied:
The LaTeX author in me says no spaces, but I don't have a strong preference.

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,298 @@
+        run_after=run_after,
+    )
+
+But since we typically want to schedule a run as soon as the data interval ends,
```
```suggestion
However, since we typically want to schedule a run as soon as the data interval ends,
```

Or just start with "Since"?

jedcunningham commented on docs/apache-airflow/howto/timetable.rst:
```diff
@@ -0,0 +1,298 @@
+.. seealso::
+
+    Module :mod:`airflow.timetables.base`
+        The public interface is heavily documented to explain what they should
```
```suggestion
        The public interface is heavily documented to explain what should
```

jedcunningham commented on docs/apache-airflow/logging-monitoring/errors.rst:
```diff
@@ -52,14 +52,17 @@ Name                                    Description
 ``operator``                            Operator name of the task that failed
 ======================================= ==================================================
 
+For backward compatibility, and additional tag ``execution_date`` is also available the represent the logical date. The tag should be considered deprecated in favor of ``data_interval_start``.
```
```suggestion
For backward compatibility, an additional tag ``execution_date`` is also available representing the logical date. The tag should be considered deprecated in favor of ``data_interval_start``.
```

kaxil commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,36 @@ Cron Presets
+.. tip::
+
+    If ``schedule_interval`` is not enough to express your DAG's schedule,
+    logical date, or data interval, see :doc:`Customizing imetables </howto/timetable>`.
```
```suggestion
    logical date, or data interval, see :doc:`Customizing timetables </howto/timetable>`.
```

kaxil commented on docs/apache-airflow/faq.rst:
```diff
@@ -216,20 +216,35 @@ actually start. If this were not the case, the backfill just would not start.
+2016-02-19 becomes available. This interval between midnights of 2016-02-19 and
+2016-02-20 is called the *data interval*, and since the it represents data in
+the date of 2016-02-19, this date is thus called the run's *logical date*, or
+the date that this DAG run is executed for, thus *execution date*.
```
```suggestion
typically summarize data. So, if you want to summarize data for ``2016-02-19``, You
would do it at ``2016-02-20`` midnight ``UTC``, which would be right after all data for
``2016-02-19`` becomes available. This interval between midnights of ``2016-02-19`` and
``2016-02-20`` is called the *data interval*, and since it represents data in
the date of ``2016-02-19``, this date is thus called the run's *logical date*, or
the date that this DAG run is executed for, thus *execution date*.
```

uranusjr replied:
Not sure about making UTC inline code; added ticks around date.

uranusjr committed:
Typos and wording/styling tweaks

ashb commented on docs/apache-airflow-providers-elasticsearch/logging/index.rst:
```diff
@@ -38,7 +38,7 @@ First, to use the handler, ``airflow.cfg`` must be configured as follows:
     [elasticsearch]
     host = <host>:<port>
-    log_id_template = {dag_id}-{task_id}-{execution_date}-{try_number}
+    log_id_template = {dag_id}-{task_id}-{data_interval_start}-{try_number}
```
The logs should be against the logical/execution date, not the data interval start (as data_interval_start is not guaranteed to be unique.

ashb replied:
At least conceptually it should be against the logcial date, even if _for now_ these have the same value.

uranusjr replied:
Maybe we should use `run_id` instead, that (together with `dag_id` and `task_id`) will forever ever be unique.

ashb commented on docs/apache-airflow/best-practices.rst:
```diff
@@ -270,77 +275,73 @@ Unit tests ensure that there is no incorrect code in your DAG. You can write uni
-         assert self.dagbag.import_errors == {}
-         assert dag is not None
-         assert len(dag.tasks) == 1
+    @pytest.fixture(scope="scope")
```
```suggestion
    @pytest.fixture(scope="module")
```

I think?

uranusjr replied:
Oh yes, my bad.

ashb commented on docs/apache-airflow/best-practices.rst:
```diff
@@ -270,77 +275,73 @@ Unit tests ensure that there is no incorrect code in your DAG. You can write uni
+
+
+    def test_my_custom_oeprator_execute_no_trigger(dag_maker):
+        with dag_maker(
```
The idea behind this doc is for users to be able to test their DAGs, but the `dag_maker` fixture is only available inside the airflow tests, so this won't work.

ashb commented on docs/apache-airflow/concepts/dags.rst:
```diff
@@ -148,11 +148,20 @@ The ``schedule_interval`` argument takes any value that is a valid `Crontab <htt
+
+    If ``schedule_interval`` is not enough to express the DAG's schedule, see :doc:`Timetables </howto/timetable>`.
+
+Every time you run a DAG, you are creating a new instance of that DAG which Airflow calls a :doc:`DAG Run </dag-run>`. DAG Runs can run in parallel for the same DAG, and each has a defined data interval, which identifies the *logical* date and time range it is running for - not the *actual* time when it was started.
```
```suggestion
Every time you run a DAG, you are creating a new instance of that DAG which Airflow calls a :doc:`DAG Run </dag-run>`. DAG Runs can run in parallel for the same DAG, and each has a defined data interval, which identifies the period of data the tasks should operate on.
```

ashb replied:
```suggestion
Every time you run a DAG, you are creating a new instance of that DAG which Airflow calls a :doc:`DAG Run </dag-run>`. DAG Runs can run in parallel for the same DAG, and each has a defined data interval, which identifies the period of data the tasks should operate on.
```

ashb commented on docs/apache-airflow/concepts/operators.rst:
```diff
@@ -66,20 +66,20 @@ Jinja Templating
 
-  # The execution date as YYYY-MM-DD
+  # The start of the data interval as YYYY-MM-DD
   date = "{{ ds }}"
```
Should we use something other than `ds` here now?

uranusjr replied:
I guess we can use `{{ run_id }}`.

ashb replied:
I was wondering if we are deprecating `{{ ds }}` -- i we aren't then this can stay as it is.

uranusjr replied:
We are not, but the semantic meaning changed (it is now backed by `data_interval_start`, which is another reason why logical date is hard wired to equal to `data_interval_start` right now).

ashb commented on docs/apache-airflow/dag-run.rst:
```diff
@@ -54,17 +54,36 @@ Cron Presets
+
+All dates in Airflow are tied to the data interval concept in some way. The
+"logical date" (also called ``execution_date`` in Airflow versions prior to 2.2)
+of a DAG run, for example, denotes the start of the data interval, not when the
```
I didn't think logical date and stat date were strictly tied to each other -- couldn't a custom timetable choose to do something else here?

uranusjr replied:
Right now logical date is hard-wired to start of data interval in the data structure. Theoratically a custom timetable can set it to something else, but it’s not exposed as a part of the API (must subclass `DataInterval`) nor explained anywhere. I think it’s best to do this as long as `executable_date` still has UNIQUE constraint to avoid people doing silly things (intentionally or not).

ashb replied:
Ah yeah, fair.

Longer term I do want to remove the unqiue constraint on execution_date, but we aren't there yet.

ashb commented on docs/apache-airflow/faq.rst:
```diff
@@ -295,7 +311,7 @@ commonly attempted in ``user_defined_macros``.
         bo = BashOperator(task_id="my_task", bash_command="echo {{ my_custom_macro }}", dag=dag)
 
-This will echo "day={{ ds }}" instead of "day=2020-01-01" for a dagrun with the execution date 2020-01-01 00:00:00.
+This will echo "day={{ ds }}" instead of "day=2020-01-01" for a dagrun with ``data_interval_start`` 2020-01-01 00:00:00.
```
```suggestion
This will echo "day={{ ds }}" instead of "day=2020-01-01" for a dagrun with a ``data_interval_start`` of 2020-01-01 00:00:00.
```

ashb reviewed (changes requested):
Looks great -- just a few small changes

uranusjr committed:
Use run_id in Elasticsearch log ID template

This ensures the log ID is unique across the board, similar to the
guarantee we had prior to AIP-39 (via execution date).

uranusjr committed:
Don't use dag_maker in Best Practice docs

uranusjr committed:
Remove logical date reference in DAG concept

uranusjr committed:
Wording tweak

uranusjr committed:
Replace ti.xcom_pull() in extra link example

We don't want to expose TaskInstance(execution_date=...) anymore, but
the extra link interface still relies on execution date, so we need an
alternative.

kaxil reviewed (commented):
Worth rebasing on `main`
